ofono/src/gprs.c

```diff
@@ -820,6 +832,252 @@ static void pri_update_mms_context_settings(struct pri_context *ctx)
+{
+	if (ctx->context.proto != ap->proto)
+		return TRUE;
+
+	if (strcmp(ctx->context.apn, ap->apn))
+		return TRUE;
+
+	if (pri_str_changed(ctx->context.username, ap->username))
+		return TRUE;
+
+	if (pri_str_changed(ctx->context.password, ap->password))
+		return TRUE;
+
+	if (ap->type == OFONO_GPRS_CONTEXT_TYPE_MMS) {
+		if (pri_str_changed(ctx->message_proxy, ap->message_proxy))
+			return TRUE;
+
+		if (pri_str_changed(ctx->message_center, ap->message_center))
+			return TRUE;
+	}
+
+	return FALSE;
+}
+
+static void pri_pending_reply_ok(gpointer data)
+{
+	DBusMessage *msg = data;
+	__ofono_dbus_pending_reply(&msg, dbus_message_new_method_return(msg));
+}
+
+static void pri_pending_reply_error(gpointer data)
+{
+	DBusMessage *msg = data;
+	__ofono_dbus_pending_reply(&msg, __ofono_error_failed(msg));
+}
+
+static void pri_reset_finish(struct pri_context *ctx, gboolean ok)
+{
+	struct context_reset *pr;
+
+	if (!ctx->pending_reset)
+		return;
+
+	pr = ctx->pending_reset;
+	ctx->pending_reset = NULL;
+
+	if (ok) {
+		pri_reset_context_properties(ctx, pr->ap);
+		g_slist_free_full(pr->msgs, pri_pending_reply_ok);
+	} else {
+		g_slist_free_full(pr->msgs, pri_pending_reply_error);
+	}
+
+	__ofono_gprs_provision_free_settings(pr->settings, pr->count);
+	g_free(pr);
+}
+
+static DBusMessage *pri_reset_async(struct pri_context *ctx, DBusMessage *msg,
+			struct ofono_gprs_provision_data *settings, int count,
+				const struct ofono_gprs_provision_data *ap)
+{
+	struct context_reset *pr;
+	struct ofono_gprs_context *gc = ctx->context_driver;
+
+	if (gc == NULL)
+		return __ofono_error_failed(msg);
+
+	pr = g_new0(struct context_reset, 1);
+	pr->msgs = g_slist_append(NULL, dbus_message_ref(msg));
+	pr->settings = settings;
+	pr->count = count;
+	pr->ap = ap;
+
+	ctx->pending_reset = pr;
+
+	/* Deactivation may already be in progress */
+	if (!ctx->pending)
+		gc->driver->deactivate_primary(gc, ctx->context.cid,
+					pri_deactivate_callback, ctx);
+
+	return NULL;
+}
+
+static DBusMessage *pri_reset_properties(DBusConnection *conn,
+					DBusMessage *msg, void *data)
+{
+	struct pri_context *ctx = data;
+	struct ofono_gprs *gprs = ctx->gprs;
+	struct ofono_modem *modem = __ofono_atom_get_modem(gprs->atom);
+	struct ofono_sim *sim = __ofono_atom_find(OFONO_ATOM_TYPE_SIM, modem);
+	struct ofono_gprs_provision_data *settings;
+	DBusMessage *reply = NULL;
+	int i, count = 0;
+
+	if (sim == NULL)
+		return __ofono_error_failed(msg);
+
+	if (ctx->pending_reset) {
+		ctx->pending_reset->msgs = g_slist_append(
+			ctx->pending_reset->msgs, dbus_message_ref(msg));
+		return NULL;
+	}
+
+	if (__ofono_gprs_provision_get_settings(ofono_sim_get_mcc(sim),
+				ofono_sim_get_mnc(sim), ofono_sim_get_spn(sim),
+						 &settings, &count) == FALSE)
+		return __ofono_error_failed(msg);
+
+	for (i = 0; i < count; i++) {
+		const struct ofono_gprs_provision_data *ap = settings + i;
+		if (ap->type == ctx->type && ap_valid(ap)) {
+			if (!(ctx->active || ctx->pending) ||
```

Comment:
To me two logical operations NOT (x OR y) is easier to comprehend than three (NOT x) AND (NOT y)